### ISSUE TITLE
Parser ignores strings at an unknown at-rule prelude (thus failing at `{` and `;` there)

### ISSUE THREAD

**Issue by sylingd** [less/less.js#3147]
I used less to write some user styles for [Stylish](https://addons.mozilla.org/en-US/firefox/addon/stylish/), it have some special syntaxs, for example, this extension support `regexp("Regular expression")`. So the following codes will work well on Stylish:

```css
@-moz-document regexp("(\d{0,15})") {
	a {
		color: red;
	}
}
```

But it can not work with less.js:
![tim 20180102004110](https://user-images.githubusercontent.com/5326684/34469247-a8beaeca-ef55-11e7-9fd2-e72f76270752.png)

The less.js says:
```
ParseError: Unrecognised input. Possibly missing opening '{' on line 1, column 32:

1 @-moz-document regexp("(\d{0,15})") {
2 	a {
```

Is there a way to ignore the characters between quotation marks? Or are there other ways to solve this problem?
Thank you

**Comment by rjgotten** [comment]
You can insert literal content using a string that is 'unquoted' via the `~` operator.
What might work for you is something like:

```less
@-moz-document regexp(~'"(\d{0,15})"') {
  a {
    color : red;
  }
}
```

If that still messes with the parser, you may consider moving the expression contents into a variable first:
```less
@regexp : ~'"(\d{0,15})"';
@-moz-document regexp(@regexp) {
  a {
    color : red;
  }
}
```

**Comment by rjgotten** [comment]
@seven-phases-max 
Crap; you're right. I forgot about the rather ignorant / non-existent parsing logic for at-rules.
In that case, I wonder if we can do any better in the look-ahead code to match the opening brace of the at-rule's block. E.g. make _some_ assumptions on sanity of the structure of the value(s) that follow the at-directive, and performing some rudimentary open/close-matching for brace/bracket/quote characters on them.

**Comment by seven-phases-max** [comment]
>In that case, I wonder if Less can do any better in the look-ahead code to match the opening brace of the at-rule's block.

I think it can try to parse that `at-rule-prelude` thing as an array of tokens (at least of two types: "string" and "anything-else", paren and bracket pairs could probably be ignored) instead of just a single "whatever" token like it does now... That should do the trick though it's hard to predict what other problems may arise when you dig into.

**Comment by matthew-dean** [comment]
@seven-phases-max (or anyone)
Hey, if you're looking at parsing and dealing with permissiveness, do you think you would have time to look at #2715 & #2986? I just added [this comment](https://github.com/less/less.js/issues/2715#issuecomment-366019034).

**Comment by matthew-dean** [comment]
@rjgotten 

> In that case, I wonder if Less can do any better in the look-ahead code to match the opening brace of the at-rule's block. E.g. make some assumptions on sanity of the structure of the value(s) that follow the at-directive, and performing some rudimentary open/close-matching for brace/bracket/quote characters on them.

It sounds to me like the rules for at-rules and custom property syntax may be similar. That is, you can include almost anything for an unknown at-rule, but there are some pair-matching rules? It might make sense to see what the overlap is and normalize the parsing between them.

**Comment by rjgotten** [comment]
@matthew-dean 

Yes; that's the conclusion I arrived at as well. ^_^
There's got to be some overlap of concerns there.

**Comment by hyperfekt** [comment]
The Stylus extension (not to be confused with the Stylus language) (also used for applying userstyles, a fork of Stylish) recently added LESS support, this bug however greatly limits its use - especially since the file workaround cannot be used.
It'd be awesome if a fix for this was to be developed soonish so we can make use of LESS. :)

**Comment by tophf** [comment]
The spec [says](https://www.w3.org/TR/css-syntax-3/#consume-an-at-rule) {-token, not {-character. The parser works by consuming/ignoring tokens. The parser should read and ignore tokens until it encounters {-token, in other words: ident-token `regexp`, (-token which starts a simple block i.e. it'll consume everything (even `{` or `}` and the doublequoted string inside) until the closing )-token. Only then it'll see {-token, and then it'll consume the {} block. LESS seems to simply search for { character which is plain wrong.

**Comment by seven-phases-max** [comment]
>The parser works by consuming/ignoring tokens.

Less parser does not. I.e. you are right about how it should be in a perfect world (above I wrote [the same](https://github.com/less/less.js/issues/3147#issuecomment-366003353)) but I'm afraid how Less parser works internally has *nothing to do* with *CSS* parsing specs.

**Comment by matthew-dean** [comment]
So I thought I'd do some tests in Chrome as to what are considered valid custom properties by the CSS custom property parser:

**TL;DR - The theory of what's valid for custom properties (in Chrome anyway) mostly matches the way the spec seems to read, with Chrome seeming to allow for unmatched braces in some cases (probably error correction?).**

**Results:**
`--test: { blah; }` - valid
`--test: { blah;` - interestingly, valid
`--test: () { blah; }` - valid
`--test: () => { blah; }` - valid
`--test: () => { blah; blah; }` - valid
`--test: () => { blah; blah blah blah; }` - valid
`--test: () => { blah; blah blah blah; [ }` - not valid
`--test: () => { blah; blah blah blah; [blah] }` - valid (confirms "matching brace" spec)
`--test: () => { blah; blah blah blah; ( }` - not valid
`--test: () => { blah; blah blah blah; (blah) }` - valid
`--test: () => { blah; blah blah blah; { }` - bizarrely, valid
`--test: () => { blah; blah blah blah; {} }` - valid
`--test: () => { blah; blah blah blah; {}}}` - not valid
`--test: () => { blah; blah blah blah;';` - valid
`--test: () => { blah; blah blah blah;";` - valid
`--test: () => {
    basically anything until final semi-colon;
  even other stuff; 
  }` - valid
`--test: () => {
    basically anything until final semi-colon;
  even other stuff; // i'm serious 
  }` - not valid
`() => {
    basically anything until final semi-colon;
  even other stuff; // im serious 
  }` - valid (with apostrophe removed, which in this case, invalidated?)
`--test: () => {
    basically anything until final semi-colon;
  even other stuff; // i\'m serious 
  }` - valid (is this in the spec? escaping?)
`--test: () => {
    basically anything until final semi-colon;
  even other stuff; // i'm serious' 
  }` - valid - apostrophe seems to be treated as "matching quote"
